numpy/core/tests/test_datetime.py

```diff
@@ -2220,6 +2221,38 @@ def test_corecursive_input(self):
         assert_raises(RecursionError, obj_arr.astype, 'M8')
         assert_raises(RecursionError, obj_arr.astype, 'm8')
 
+    @pytest.mark.parametrize("op, cmp_op, inv_op", [
+        (operator.add, operator.gt, operator.sub),
+        (operator.sub, operator.lt, operator.add),
+    ])
+    @pytest.mark.parametrize("time_unit", [
+        "Y", "M",
+        "W",  # gh-13197 means this is broken but this test does not catch it
+        "D", "h", "m",
+        "s", "ms", "us", "ns", "ps", "fs", "as",
+    ])
+    def test_limit_symmetry(self, time_unit, op, cmp_op, inv_op):
+        """
+        Dates should have symmetric limits around the unix epoch at +/-np.int64
+        """
+        # TODO: add absolute (gold standard) time span limits
+        epoch = np.datetime64(0, time_unit)
+        max_delta = np.timedelta64(np.iinfo(np.int64).max, time_unit)
+
+        # subtraction is particularly prone to overflow in older code, making
+        # this fail
+        limit = op(epoch, max_delta)
+        assert cmp_op(limit, epoch)
+
+        # The overflow would show up in the repr so test a roundtrip from epoch
+        # to limit and back using the str representation
+        limit_via_str = np.datetime64(str(limit))
+        assert limit_via_str == limit
+
+        # finish the roundtrip to unix epoch:
+        restored = inv_op(limit, max_delta)
+        assert restored == epoch
```

Comment:
This isn't as sensitive as what I had before. I can restore the sensitivity to issues with weeks:

```diff
--- a/numpy/core/tests/test_datetime.py
+++ b/numpy/core/tests/test_datetime.py
@@ -2237,20 +2237,15 @@ class TestDateTime(object):
         """
         # TODO: add absolute (gold standard) time span limits
         epoch = np.datetime64(0, time_unit)
-        max_delta = np.timedelta64(np.iinfo(np.int64).max, time_unit)
-
-        # subtraction is particularly prone to overflow in older code, making
-        # this fail
-        limit = op(epoch, max_delta)
-        assert cmp_op(limit, epoch)
+        max_delta = np.timedelta64(np.iinfo(np.int64).max)
 
         # The overflow would show up in the repr so test a roundtrip from epoch
         # to limit and back using the str representation
-        limit_via_str = np.datetime64(str(limit))
-        assert limit_via_str == limit
+        limit_via_str = np.datetime64(str(op(epoch, max_delta)))
+        assert cmp_op(limit_via_str, epoch)
 
         # finish the roundtrip to unix epoch:
-        restored = inv_op(limit, max_delta)
+        restored = inv_op(limit_via_str, max_delta)
         assert restored == epoch

```

I can understand some desire for being explicit with the `m8` units, but it should be perfectly valid to add generic deltas to datetime64 values in the respective units from each parametrization. Being implicit increases sensitivity if you also use the string path on the final assert.

Reply:
Something isn't making sense here. I'll have a play around with the union of our two tests

Reply:
```python
ref = np.datetime64(0, 'W')
max_delta = np.timedelta64(np.iinfo(np.int64).max)
max_delta_units = np.timedelta64(np.iinfo(np.int64).max, 'W')
generic_round_trip = np.datetime64(np.datetime64(str(ref - max_delta)) + max_delta) 
units_round_trip = np.datetime64(np.datetime64(str(ref - max_delta_units)) + max_delta_units)
print(generic_round_trip) # 1970-01-07
print(units_round_trip) #1970-01-01
```

Reply:
Thanks for the easily-pasteable example

That test isn't valid, and produces different results without overflow:
```
ref = np.datetime64(0, 'W')
delta = np.timedelta64(1)
delta_unit = np.timedelta64(1, 'W')
ref_plus_delta = ref + delta
assert ref_plus_delta == ref + delta_unit  # makes no difference here

# this converts from `M8[W]` to `M8[D]
ref_plus_delta_str = np.datetime64(str(ref_plus_delta))

r1 = ref_plus_delta_str - delta_unit  # subtracts weeks
r2 = ref_plus_delta_str - delta  # subtracts days, since the unit changed
```

What you've proven there is that:
* generic timedeltas are fragile
* Round-tripping via string might change the unit

Reply:
I still think the test was just fine as it was--unless those fragilities / behaviors are intentional it seems like enforcing the expectation that those tests pass is reasonable, with a skip on the weeks for now.

If it is just weeks that magically switches like that it seems like something to be addressed for weeks and locked down for the other units in the test.

Reply:
Maybe even add another assertion on the `m8` unit of time preservation through the string conversion.

Reply:
> Maybe even add another assertion on the m8 unit of time preservation through the string conversion.

Sounds like a good idea for a separate test.

> unless those fragilities / behaviors are intentional

They are. There's no way to tell from `1970-01-01` if the user wanted `W`, `D`, `2D`, etc - we choose a unit based on the precision of the date string, and we have no prevision between `YYYY-MM` and `YYYY-MM-DD`

Reply:
If you convert from weeks to days near the integral size limit you have an impossible decision to make, right? What year is it? We can't balance the magnitude of the date with the precision of the unit. I can't think of a sensible reason we should allow that or how it would behave logically.

We could probably agree that the unit test for round tripping symmetry is just tying to do too much if you throw in the generic `m8` issues as well. It may be rather verbose to add a second parametrized test over all the units for that, but fine by me if you prefer that.

Reply:
> I can't think of a sensible reason we should allow that or how it would behave logically.

In my opinion, `np.datetime64("ONLY_FITS_IN_WEEKS-01-01")` should error / give NaT, not retry with weeks if days do not fit.

Remember the user can always ask for  `np.datetime64("ONLY_FITS_IN_WEEKS-01-01", 'W')`